lib/course_planner/notifier.ex

```diff
@@ -43,8 +39,35 @@ defmodule CoursePlanner.Notifier do
         {:noreply, [{:error, reason, email} | state]}
     end
   end
+  def handle_cast({:save_email, notification}, state) do
+    changeset = Notification.changeset(notification)
+    case Repo.insert(changeset) do
+      {:ok, _} ->
+        {:noreply, state}
+      {:error, %{errors: errors, data: email}} ->
+        Logger.error("Email saving failed: #{Kernel.inspect errors}")
+        {:noreply, [{:error, errors, email} | state]}
+    end
+  end
+  def handle_cast({:notify_all, user}, state) do
+    email = UserEmail.build_summary(user)
+    case Mailer.deliver(email) do
+      {:ok, _} ->
+        delete_notifications(user)
+        {:noreply, state}
+      {:error, reason} ->
+        Logger.error("Email delivery failed: #{Kernel.inspect reason}")
+        {:noreply, [{:error, reason, email} | state]}
+    end
+  end
   def handle_cast(_, state), do: {:noreply, state}
 
+  defp delete_notifications(user) do
```

Comment:
this implementation can delete notifications which were not sent yet if a notification is created between the notify_all casting and the delete notifications.
what do you think of mapping all the notification ids which are loaded in the user already and then delete by notification id?